app/services/bulk/bulk_upload.rb

```diff
@@ -19,8 +19,11 @@ def process
             secondary_address = table.by_row[i]["secondary_address"]
             @user = User.new(username: username , email: email)
             if @user.save
-            @user_detail = UserDetail.create(first_name: first_name , last_name: last_name , primary_address: primary_address , secondary_address: secondary_address , dob: dob , user_id: @user.id)
-            else 
+            	@user_detail = UserDetail.create(first_name: first_name , last_name: last_name , primary_address: primary_address , secondary_address: secondary_address , dob: dob , user_id: @user.id)
+						 if @user_detail.errors.full_messages_for(:dob).length == 1
+								@user.destroy
```

Comment:
Try the accepts_nested_attributes_for method to save nested data in one go. If any validation fails it will automatically roll back. No need to destroy the user seperately.